src/background/StreamSessionManager.ts

```diff
@@ -0,0 +1,373 @@
+    }
+
+    /**
+     * Releases managed resources
+     */
+    public destroy(): void {
+        this.peer.destroy();
+    }
+
+    /**
+     * Handle a new peer connection
+     * @param conn The underlying data connection
+     * @param userDisplayName The name of the user who initiated the connection
+     */
+    private handleConnect(conn: Peer.DataConnection, userDisplayName: string): void {
+        console.warn("handleConnect");
+        // Add new connection to local users
+        const user: IUserConnection = {
+            id: conn.peer,
+            displayName: userDisplayName,
+            connection: conn,
+            permissions: {
+                canControlVideo: true
+            }
+        };
+        this.sessionState.sessionUsers.set(user.id, user);
+
+        conn.on("data", (data: Message) => {
+            console.warn(`data from: ${conn.peer}`);
+            console.warn(data);
+            switch (data.messageType) {
+                case MessageType.StateSync:
+                    this.handleStateSyncMessage(data);
+                    break;
+                case MessageType.StateSyncRequest:
+                    this.handleStateSyncRequestMessage();
+                    break;
+                default:
+                    break;
+            }
+
+            this.port.postMessage(data);
+        });
+
+        conn.on("close", () => {
+            this.sessionState.sessionUsers.delete(user.id);
+            this.sessionState.sessionJoinOrder.splice(this.sessionState.sessionJoinOrder.indexOf(user.id), 1);
+            this.notifyUserCountChanges(-1, false);
+            if (user.id === this.sessionState.leaderId) {
+                const newLeaderId = this.determineNewLeaderId();
+                this.sessionState = { ...this.sessionState, leaderId: newLeaderId };
+                if (newLeaderId === this.currentUserId) {
+                    // If the current user is the new leader, they should do a full sync after a processing wait
+                    setTimeout(() => {
+                        const syncStateMessage = this.generateSyncStateMessageFromCurrentState();
+                        this.sessionState.sessionUsers.forEach(user => {
+                            if (user.id !== this.currentUserId) {
+                                user.connection?.send(syncStateMessage);
+                            }
+                        });
+                    }, 500);
+                }
+            }
+        });
+    }
+
+    /**
+     * Determines the new leader id after the leader disconnects. Assumes the previous leader has been removed from sessionState
+     */
+    private determineNewLeaderId(): string {
+        for (const userId of this.sessionState.sessionJoinOrder) {
+            const user = this.sessionState.sessionUsers.get(userId);
+            if (user?.permissions.canControlVideo) {
+                return user.id;
+            }
+        }
```

Comment:
Could this just be...

```ts
const nextLeader =  this.sessionState.sessionUsers.entries().filter((user) => user.permissions.canControlVideo);

if (nextLeader) {
  return nextLeader.id;
}

return this.sessionState.sessionUsers.entries()[0];
```

I know `Set()` doesn't preserve order ... but .... 🤷‍♀️ it kinda does.

Reply:
That doesn't work because we want the first one in the join order with canControlVideo perms to be the next leader. If no one else has those perms then we go with the next up in the join order array